### ISSUE TITLE
Unused parameters

### ISSUE THREAD

**Issue by MikeGitb** [cinder/Cinder#1629]
## Summary

On higher warning levels, compiling Cinder produces 100s of warning related to unused function parameters and I'd like to get rid of them. This issue is supposed to serve as a central point to discuss the general problem and specific cases, where the bes course of action is not clear.
## Explanation

When compiling with higher warning levels (e.g. /W4 on VS or -Wall on g++/clang) the compiler warns about unused parameters. In general, this is a good thing, as it might warn you that you e.g. forgot to implement a certain functionality, are using a local / member variable with a similar name instead of the parameter, have some other bug or simply that you should rethink the interface.

However, when I'm compiling the library from the master branch with VS2015 U3, I currently get around 680 of those warnings and most of them are harmless, which is a tremendous amount of noise. Even worse: A lot of those warnings alsow appear to the users of cinder when they include the header files (Just compiling the BasicApp example I results in 12 of them). 

If there is interest in this, I'm intending to resolve most of those warnings:
- In most cases, those parameters aren't used by design, e.g. because those functions implement a general interface, but the particular implementation doesn't need some of the parameters. In those cases, I'd just comment out the parameter names:
  
  ```
  int foo_old(int i) {  return42; }
  int foo_new(int /*i*/) { return 42; }
  ```
  
  This will never introduce a new bug, but it might "hide" a bug if ignoring the parameter wasn't actually intended.
- In some cases the function in question should really be a pure abstract function instead:
  
  ```
  virtual void foo_old(int i) {}
  virtual void foo_old(int i) = 0;
  ```
  
  In addition to removing the warning, this also makes the intent more clear and derived classes can not forget to override a function. This will however (obviously) cause build errors, if the function was actually meant to be a do-nothing-default-implementation.
- Some cases will be actual bugs (e.g. see #1628) and I can either offer a PR or raise a new issue
- Unclear cases could be discussed here

Priority wise I'd first target any warnings coming from header files, as those will create new warnings for every translation unit and also leak to the users of cinder.

**Comment by andrewfb** [comment]
I'd say on the whole we're definitely interested in fixing these, and as you said, the highest priority warnings would be those that find their way into user apps. Frankly, we have a number of warnings that are triggered at lower warning levels too. In the constant triage of work on Cinder, these have not found their way to the top of the priority queue, but we'd definitely welcome this kind of contribution.

There's probably a balance to be struck between too many and too few warnings in a single PR, so if there's some sort of natural categorization that would help accelerate us vetting and merging any PR's.

Thanks for your willingness and interest in tackling these.

**Comment by MikeGitb** [comment]
@andrewfb: Yes, you are right about prioritizing warnings on lower levels and 'm also looking into some of them (#1599). But the unused parameter warnings can (mostly) be solved with relatively simple modifications and little understanding of the code base. So for a beginner like me it seemes like a good way to get around the code base and at least do a little good when stumbling over actual bugs on the way.

**Comment by richardeakin** [comment]
I'd definitely like to see a reduction of warnings when building in VS and haven't been able to contribute to that myself. I recall we did a large pass on the OS X side for 0.9.0, but some remained that required extra thought for another day (IIRC they were situations where pointer arithmetic was going on).

I would suggest to group the modifications into PRs pertaining to one area of the code, for example the `ci::app` layer, or Geoms, etc. That way we can review the code in a timely manner without it being too overwhelming.

**Comment by MikeGitb** [comment]
@richardeakin: I'll see what I can do. A lot of warnings are about narrowing conversions though and I'm unsure whether I should simply silence them by introducing static_casts. "Clean" resolutions (where possible at all) will most likely require interface changes, which I don't feel qualified to make yet or checked narrowing conversions which would introduce a lot of overhead.

**Comment by paulhoux** [comment]
Thanks for creating this issue. The warnings have "bugged" me for a while as well, as it isn't a sign of a clean code base. So far, I have not attempted to fix them myself, mainly because with different people working on the code base at the same time it would cause a lot of merge conflicts. But these can be resolved as well, of course.

**Comment by MikeGitb** [comment]
I think there is something off at [`cinder::audio::TargetFile`](https://github.com/cinder/Cinder/blob/master/include/cinder/audio/Target.h#L50-L52) The constructor takes a `DataTargetRef` parameter, but ignores it. I haven't enought understanding of the usacases of that class to determine if this is just an error (although I don't see, what the parameter should be used for) or a legacy artifact. And if it is the latter, whether the parameter can be removed completely or if that would break some exeisting code.

**Comment by MikeGitb** [comment]
I think the `multisampling` parameter [here](https://github.com/cinder/Cinder/blob/master/src/cinder/gl/Fbo.cpp#L308) can safely be removed, but it could also be a bug / missing feature.

**Comment by MikeGitb** [comment]
I dealt with almost all wanna related to unused parameters and will soon start to package them into PRs. Should I create them against master or the android linux branch?

**Comment by MikeGitb** [comment]
The fact that the [parameter b of makeWeightTable in Resize.cpp](https://github.com/cinder/Cinder/blob/master/src/cinder/ip/Resize.cpp#L232) isn't used seems to be either a bug, or a reason to remove it, as it is used locally anyway. However, as I don't know, what it is supposed to do I can't say which is which.

**Comment by richardeakin** [comment]
Thanks! I think if you can make them grouped nicely and they don't affect new files instroduced by android and / or linux, PR'ing against master would be great. I'll get it merged into android_linux after.

**Comment by richardeakin** [comment]
Carrying on from the discussion in #1671, I think the 'policy' we arrived at for removing unused warnings from public API methods was around the following:

- We definitely want to silence the warnings.
- If the variable shouldn't be there in the first place, let's go ahead and remove it rather than comment it out where it will go forgotten
- In some cases, like in virtual methods with a default empty implementation, consider moving the default implementation to the cpp, where something like

```
void Node::process( Buffer */*buffer*/ )
{
}
```

won't make the public API more difficult to read.

- Otherwise, using `/* */` to remove the variable in headers is a better option over something like an `unused( var )` macro.

We're obviously open to discussion more around these policies, and eventually we should probably get them into the CONTRIBUTING.md guide. With that, I think we're in a good spot to move forward with getting #1669, #1670 and #1671 merged.

**Comment by MikeGitb** [comment]
Great!

> If the variable shouldn't be there in the first place, let's go ahead and remove it rather than comment it out where it will go forgotten

I agree. Unfortunately it was not always obvious for me which is which, so feel free to adapt the PRs accordingly. When you review / merge the PRs please alsoo keep in mind that some instances of unused parameters might be an intication of an actual error not just due to unsupported functionality. I hope I addressed all of those cases in previous PRs or the comments above, but one or two additional pairs of eyes cerrtainly wouldn't hurt.

**Comment by MikeGitb** [comment]
Oh and if you have time, you might also want to have a look at already merged PRs as they don't quite follow the policy you mentioned here.

**Comment by richardeakin** [comment]
Thank again @MikeGitb, these are definitely bring up some spots here and there that we should be thinking about how to improve, and it has been great how they've been kept modular so that we can separately consider those cases and gradually get the improvements merged.

**Comment by MikeGitb** [comment]
Just to wrap this up, here is a list of the remaining warnings about unused parameters that I get when compiling cinder:
- [src/cinder/ip/Resize.cpp](https://github.com/cinder/Cinder/blob/master/src/cinder/ip/Resize.cpp#L232): makeWeightTable doesn't use the first parameter `b`. 
 -> No Idea if this intentional or a bug
- [src/cinder/audio/msw/FileMediaFoundation.cpp](https://github.com/cinder/Cinder/blob/master/src/cinder/audio/msw/FileMediaFoundation.cpp#L409): `TargetFileMediaFoundation` constructor doesn't use the `extension` parameter
 -> Can probably be removed
- [src/cinder/audio/Target.cpp#L42] (https://github.com/cinder/Cinder/blob/master/src/cinder/audio/Target.cpp#L42): The two `TargetFile::create` functions don't use the `extension parameter`
 -> Can probably be removed
- [src/cinder/Font.cpp](https://github.com/cinder/Cinder/blob/master/src/cinder/Font.cpp#L300): The callback `EnumFontFamiliesExProc` doesn't use `lpntme` and `FontType`
 -> Probably has to conform to an expected function signature, so I'd comment out the parameter names

**Comment by richardeakin** [comment]
About the `extension` param for `audio::Target` and `audio::msw::FileMediaFoundation`, it will be in use once https://github.com/cinder/Cinder/pull/1869 is merged.

I have no idea on the `ip::Resize` one, @andrewfb should know. As far as `Font` goes, I think you're right and agree with your fix.

**Comment by MikeGitb** [comment]
As there are at least potential fixes for the warnings explicitly mentioned in this issue (most likely new ones have crept up on the mean time) I'm closing this issue.
